### PR TITLE
Update event connection order

### PR DESCRIPTION
**Pull request by Czaki**
# References and relevant issues

In PartSeg I got multiple such reports:

```python
ValueError: Volume visual needs a 3D array.
  File "PartSeg\common_gui\napari_image_view.py", line 682, in _add_image
  File "PartSeg\common_gui\napari_image_view.py", line 645, in _add_layer_util
  File "napari\components\viewer_model.py", line 742, in add_layer
  File "_collections_abc.py", line 1128, in append
  File "napari\components\layerlist.py", line 188, in insert
  File "napari\utils\events\containers\_selectable_list.py", line 68, in insert
  File "napari\utils\events\containers\_evented_list.py", line 202, in insert
  File "napari\utils\events\event.py", line 764, in __call__
  File "napari\utils\events\event.py", line 802, in _invoke_callback
  File "napari\utils\events\event.py", line 789, in _invoke_callback
  File "napari\_qt\qt_viewer.py", line 658, in _on_add_layer_change
  File "napari\_qt\qt_viewer.py", line 668, in _add_layer
  File "napari\_vispy\utils\visual.py", line 88, in create_vispy_layer
  File "napari\_vispy\layers\image.py", line 89, in __init__
  File "napari\_vispy\layers\scalar_field.py", line 71, in __init__
  File "napari\_vispy\layers\scalar_field.py", line 123, in _on_data_change
  File "vispy\visuals\volume.py", line 852, in set_data
  ```

# Description

When try to debug the above error, I have found that the callback order looks incorrect 

obraz

so `QtViewer._on_add_layer_change` is called before `Viewer._on_add_layer` when the `Viewer._on_add_layer` is triggering data slicer, so  data slice value may be incorrect. 

I think that after #4480 we should keep events in connection order. So Qt updates will be triggered after core models updates 

This PR also updates the code for fixing data shape  to work good with data of `ndim=1`

This PR is a workaround problem that is reported to me as it happens on Windows, on long living sessions.

### CODE REVIEW

File: napari/utils/events/event.py [hunk 1]
```diff
@@ -410,7 +410,7 @@ def connect(
         self,
         callback: Union[Callback, CallbackRef, CallbackStr, 'EventEmitter'],
         ref: Union[bool, str] = False,
-        position: Literal['first', 'last'] = 'first',
+        position: Literal['first', 'last'] = 'last',
```
Comment (brisvag) [inline napari/utils/events/event.py:413]: This is a pretty big change right? It means every event order in the whole codebase and plugins will be swapped? Are we sure we want to change the default here?

Comment (Czaki) [reply napari/utils/events/event.py:413]: Yes. After this change, events will be executed in connection order, not in reversed order. So callbacks from ViewerModel will be executed before callbacks from QtViewer

Comment (brisvag) [reply napari/utils/events/event.py:413]: I think in principle this is how I would expect it to work, but it's likely to affect things... I'm surprised tests are all passing xD

Comment (Czaki) [reply napari/utils/events/event.py:413]: Because the callbacks should be independent. 
But my investigation in original bug show that sometimes, during processing of callbacks, some QtLoop event processing happens and reval some problem. 

Also reported problem may happen only when using 3d view.

Comment (brisvag) [reply napari/utils/events/event.py:413]: Ok. I would ask some folks with complex plugins to try out this change before merging! I'll try with blik later.

Comment (melonora) [reply napari/utils/events/event.py:413]: good point, I will try with napari-spatialdata. One thing @Czaki, correct me if I am wrong, this does not affect the order of events being emitted by QtViewer right (asking because of differences between windows / linux and macs here)? Just solely the viewer model first and then the QtViewer.

Comment (Czaki) [reply napari/utils/events/event.py:413]: `QtViewer` events are Qt events, so there should be no impact.

Comment (brisvag) [reply napari/utils/events/event.py:413]: > Ok. I would ask some folks with complex plugins to try out this change before merging! I'll try with blik later.

All seems to be working well on my side!

File: napari/_vispy/layers/scalar_field.py [hunk 2]
```diff
@@ -103,8 +103,8 @@ def _on_data_change(self) -> None:
             ndisplay, getattr(data, 'dtype', None)
         )
 
-        if ndisplay == 3 and self.layer.ndim == 2:
-            data = np.expand_dims(data, axis=0)
+        if ndisplay > data.ndim:
+            data = data.reshape((1,) * (ndisplay - data.ndim) + data.shape)
```
Comment (jni) [inline napari/_vispy/layers/scalar_field.py:107]: Should this be in a different PR?

**Comment by jni** [comment]
I like the change but I feel like this was the *intent* with #4480? Why did we default to 'first' then?

**Review by jni — APPROVED** [review]

File: napari/_vispy/layers/scalar_field.py (hunk 2 above)
Comment (jni) [reply napari/_vispy/layers/scalar_field.py:107]: Other than this change I think this is ready to merge. But I would like to keep independent fixes independent.

**Comment by Czaki** [comment]
> I like the change but I feel like this was the _intent_ with #4480? Why did we default to 'first' then?

The #4480 moved to first half only napari events (so electively create two callback queue). The `first` was set in #158, the original PR introducing which mechanism,

**Review by brisvag — APPROVED** [review]
LGTM! Yeah maybe moving the other fix out is a good idea.
